### PR TITLE
config: Add `envoy.features.disallow_deprecated_features` runtime key

### PR DESCRIPTION
**Pull request by auni53**
Commit Message:
* Enabling this flag causes deprecated field usage to crash instead of warning
* This matches the behaviour of -DENVOY_DISABLE_DEPRECATED_FEATURES, because bazel doesn't allow transitive compile-time flags.

Risk Level: Low
Testing: Basic unit tests

Signed-off-by: Auni Ahsan <auni@google.com>

### CODE REVIEW

**Comment by auni53** [comment]
1. Open to whatever renaming might be necessary.
2. Are there any other places I need to add documentation? Did my best to infer from similar previous PRs.
3. Is there any worry about using `Runtime::runtimeFeatureEnabled` vs `runtime->snapshot().featureEnabled`?

**Comment by auni53** [comment]
Hmm, I'm failing format tests but I see no output from `> ./ci/run_envoy_docker.sh 'ci/check_and_fix_format.sh'`, and the artifact produced by the presubmit has an empty diff.

**Comment by asraa** [comment]
Ugh, that's kinda annoying. This might be the error:
```
ERROR: ./docs/root/version_history/current.rst:44: Version history line malformed. Does not match VERSION_HISTORY_NEW_LINE_REGEX in check_format.py
 * config: Add `envoy.features.disallow_deprecated_features` runtime flag, which matches the behaviour of compile-time flag `ENVOY_DISABLE_DEPRECATED_FEATURES`, i.e. use of deprecated fields will cause a crash.
ERROR: check format failed. run 'tools/code_format/check_format.py fix'
```

It looks like it's not alphabetical, hopefully that's the problem?

**Comment by asraa** [comment]
I think it wants a lowercase after the colon :/ (I would even opt for a comment above that line with an example because I hate reading regexi)
https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/tools/code_format/check_format.py#L153

**Review by asraa — COMMENTED** [review]
Couple of comments:
* Should this also be changed if the runtime feature is enabled? https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/extensions/common/utility.h#L29
* Maybe a bullet point should be added here? https://github.com/envoyproxy/envoy/blob/17e815122ff53d0ac6cb2d64cdbf1bfc547bb7e8/CONTRIBUTING.md#breaking-change-policy 
* And here? https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/docs/root/configuration/operations/runtime.rst#using-runtime-overrides-for-deprecated-features

Otherwise, change looks good to me
@alyssawilk do you think those are right ^?

File: source/common/runtime/runtime_features.cc
Comment (alyssawilk) [inline source/common/runtime/runtime_features.cc:106]: So runtime guards added here are intended to be flipped true and removed.
Do you want a regular runtime guard rather than a feature with a finite lifetime?

File: tools/code_format/check_format.py
Comment (alyssawilk) [inline tools/code_format/check_format.py:153]: Would it make more sense to update the error message to explain better?

**Review by alyssawilk — COMMENTED** [review]
Cool, I think this will be a useful feature, thanks for adding it!

File: test/common/protobuf/utility_test.cc
Comment (htuch) [inline test/common/protobuf/utility_test.cc:1574]: Does this need to be a `DEPRECATED_FEATURE_TEST`? I think it would run time in the fatal-by-default build mode (since you are basically doing this at runtime).

File: source/common/protobuf/utility.cc
Comment (htuch) [inline source/common/protobuf/utility.cc:194]: Would it make sense to use the supplied `runtime` argument to this method? Not sure if that helps with testability etc, but easier to reason about one runtime instead of a global and local one.

**Review by htuch — COMMENTED** [review]
LGTM modulo various comments.

File: source/common/protobuf/utility.cc
Comment (auni53) [reply source/common/protobuf/utility.cc]: Done.

File: source/common/runtime/runtime_features.cc
Comment (auni53) [reply source/common/runtime/runtime_features.cc]: I'm not sure if I understand what you mean by regular runtime guard, but I interpreted this as moving it into the runtime features allowlist above and changed the flag name so it could default to true.

File: test/common/protobuf/utility_test.cc
Comment (auni53) [reply test/common/protobuf/utility_test.cc]: Removed.

File: tools/code_format/check_format.py
Comment (auni53) [reply tools/code_format/check_format.py:153]: I'm not really familiar with any of this infrastructure, just added this comment after tripping on the presubmit, so I'd rather not fix the error message here.

File: source/common/runtime/runtime_features.cc
Comment (auni53) [reply source/common/runtime/runtime_features.cc]: Re-interpreted this after seeing https://github.com/envoyproxy/envoy/pull/13607/files

Removed from the features list.

**Review by auni53 — COMMENTED** [review]
@asraa I added comments to the two docs you listed, thanks for finding those. I think this should not affect [deprecatedExtensionStatus](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/extensions/common/utility.h#L29) as we'd increment the stat regardless of whether we warn or crash.

File: docs/root/configuration/operations/runtime.rst
Comment (htuch) [inline docs/root/configuration/operations/runtime.rst:259]: Inconsistent name.

File: CONTRIBUTING.md
Comment (htuch) [inline CONTRIBUTING.md:76]: I think the name is a bit confusing given we already have `envoy.features.enable_all_deprecates_features`. How about `envoy.features.fail_on_any_deprecated_feature` with reversed logic?

Comment (alyssawilk) [reply CONTRIBUTING.md:76]: runtime flag -> runtime key here and below, please

File: tools/code_format/check_format.py
Comment (alyssawilk) [reply tools/code_format/check_format.py:153]: sure, but if you're adding a comment, everyone after you is going to deal with the same confusing error message.  Better to change that?

        if not match:
          reportError("Version history line malformed. "
                      "Does not match VERSION_HISTORY_NEW_LINE_REGEX in check_format.py\n %s" %
                      line)

You could change the text to something like
 
Does not match VERSION_HISTORY_NEW_LINE_REGEX in check_format.py\n %s\nPlease use messages in the form 'category: feature explanation, starting with lower-cased letter and ending with a period.'"

or something?

**Review by alyssawilk — APPROVED** [review]

**Comment by alyssawilk** [comment]
@dio can you do the non-google pass?

**Review by htuch — DISMISSED** [review]
LGTM, defer to @dio

**Comment by auni53** [comment]
Can this be merged?

**Comment by htuch** [comment]
@auni53 you need to get format to pass. @dio can you do the non-Googler pass?

**Comment by asraa** [comment]
@snowp could you take a quick look to merge please? I think Dio is out.

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
